libraries/helpers.rb

```diff
@@ -23,12 +23,17 @@ module Cookbook
     module Helpers
       include Utils
 
-      def installed_postgresql_major_version
+      def installed_postgresql_package_logic
```

Comment:
Let's refactor this so it's a little clearer:
```ruby
def installed_postgresql_package_logic
  pgsql_package = node['packages'].filter { |p| p.match?(/^postgresql-?(\d+)?$/) }.values

  unless pgsql_package.one?
    Chef::Log.warn("Detected #{pgsql_package.count} installed PostgreSQL versions: #{pgsql_package.keys.join(', ')}. Using latest version.")
    pgsql_package = pgsql_package.max_by { |_key, value| value['version'].to_f }
  end

  raise 'Unable to determine installed PostgreSQL version' if nil_or_empty?(pgsql_package)

  pgsql_package.first
end
```

Reply:
I was able to implement your refactor/changes with a few modifications

- The Chef Log.warn had to be modified as the '.keys' call no longer existed after extracting the values in line 1. I then modified this to a var which maps over all versions and pulls them together into a single string.
- The '_key' and 'value' separation of the sort had to be reduced to just 'values' as the key no longer exists after calling .values in the first line.
- The 'pgsql_package.max_by' call had to be wrapped in an array to ensure the first value out of the hash is not pulled i.e 'epoch'.

`def installed_postgresql_package_logic
        pgsql_package = node['packages'].filter { |p| p.match?(/^postgresql-?(\d+)?$/) }.values

        unless pgsql_package.one?
          versions = pgsql_package.map { |values| values['version'] }.join(', ')
          Chef::Log.warn("Detected #{pgsql_package.count} installed PostgreSQL versions: #{versions}. Using latest version.")
          pgsql_package = [pgsql_package.max_by { |values| values['version'].to_f }]
        end

        raise 'Unable to determine installed PostgreSQL version' if nil_or_empty?(pgsql_package)

        pgsql_package.first
      end`
      
  Let me know if there's anything else you want modified or more obvious

Reply:
LGTM!